src/mplex.js

```diff
@@ -0,0 +1,313 @@
+  reset (err) {
+    err = err || new Error('Underlying stream has been closed')
+    this._chandata.end(err)
+    this.close(err)
+  }
+
+  /**
+   * Pushes data to the stream
+   * @param {Buffer} data
+   */
+  push (data) {
+    if (data.data &&
+      Buffer.byteLength(data.data) > this._maxMsgSize) {
+      this._chandata.end(new Error('message too large!'))
+    }
+
+    this._chandata.push(data)
+  }
+
+  /**
+   * Creates a new Channel (stream). If Mplex was created with `opts.lazy` set to true,
+   * the channel will not automatically be opened.
+   *
+   * @param {string} name The name of the channel/stream to create
+   * @returns {Channel}
+   */
+  createStream (name) {
+    if (typeof name === 'number') { name = name.toString() }
+    const chan = this._newStream(null, true, false, name, this._outChannels)
+    if (!this._lazy) { chan.openChan(name) }
+    return chan
+  }
+
+  /**
+   * Attempts to create a channel if it doesn't already exist.
+   * If a channel already exists for `id`, `Mplex#error` will be emitted.
+   *
+   * @private
+   * @param {number} id The id of the channel. If `null` it will be auto incremented from `Mplex._chanId`
+   * @param {boolean} initiator Is the channel creating the connection
+   * @param {boolean} open Should the channel be opened when created
+   * @param {string} name The name of the channel
+   * @param {Array} list The channel list to add the channel to
+   * @returns {Channel}
+   */
+  _newStream (id, initiator, open, name, list) {
+    if (this.chanSize >= this._maxChannels) {
+      this.emit('error', new Error('max channels exceeded'))
+      return
+    }
+
+    if (typeof initiator === 'string') {
+      name = initiator
+      initiator = false
+      open = false
+    }
+
+    if (typeof open === 'string') {
+      name = open
+      open = false
+    }
+
+    id = typeof id === 'number' ? id : this._chanId++
+    if (list[id]) {
+      this.emit('error', new Error(`channel with id ${id} already exist!`))
+      return
+    }
+    const chan = new Channel({
+      id,
+      name,
+      plex: this,
+      initiator,
+      open: open || false
+    })
+
+    return this._addChan(id, chan, list)
+  }
+
+  _addChan (id, chan, list) {
+    chan.once('close', () => {
+      list[id] = null
+    })
+    chan.once('error', (err) => {
+      log.err('channel error', err)
+    })
+
+    list[id] = chan
+    return chan
+  }
+
+  get chanSize () {
+    return this._inChannels.size + this._outChannels.size
+  }
+
+  /**
+   * Takes the appropriate course of action based on `msg.type`.
+   * If `msg.type` is not recognized `Mplex#error` will be emitted.
+   * @param {Message} msg
+   */
+  _handle (msg) {
+    this._log('_handle', msg)
+    const { id, type, data } = msg
+    switch (type) {
+      // Create a new stream
+      case Types.NEW: {
+        const chan = this._newStream(id, false, true, data.toString(), this._inChannels)
+        nextTick(() => this.emit('stream', chan, id))
```

Comment:
You can avoid using a closure here by using a top-level function and passing in arguments, like so: `nextTick(emitStream, that, chan, id)`.